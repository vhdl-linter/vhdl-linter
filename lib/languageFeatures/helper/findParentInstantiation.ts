import { Position } from "vscode-languageserver";
import { OAssociationList, OFile, OInstantiation } from "../../parser/objects";
import { VhdlLinter } from "../../vhdl-linter";
import { findObjectFromPosition } from "../findObjects";

export function findParentInstantiation(linter: VhdlLinter, position: Position): [OInstantiation, OAssociationList | undefined] | undefined {
  const object = findObjectFromPosition(linter, position)[0];
  if (object === undefined) {
    return undefined;
  }
  let iterator = object;
  let associationList: OAssociationList | undefined;
  // Find Parent that is defined by a subprogram (instantiation)
  while (iterator instanceof OFile === false) {
    if (iterator instanceof OAssociationList) {
      associationList = iterator;
    }
    if (iterator instanceof OInstantiation) {
      return [iterator, associationList];
    }

    if (iterator.parent instanceof OFile) {
      break;
    }
    iterator = iterator.parent;
  }
  return undefined;
}
