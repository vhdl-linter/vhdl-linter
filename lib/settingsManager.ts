import { connection, documents, validateTextDocument } from "./languageServer";
import { defaultSettings, ISettings, normalizeSettings } from "./settings";

let globalSettings: ISettings = defaultSettings;
export const currentCapabilities = {
  configuration: false,
  workspaceFolder: false
};
// Cache the settings of all open documents
export const documentSettings = new Map<string, ISettings>();

export async function getDocumentSettings(resource?: URL): Promise<ISettings> {
  // default settings are assumed as default and the overwritten by either
  // settings from vs code (workspace) or the closest vhdl-linter.yml
  if (!currentCapabilities.configuration) {
    return normalizeSettings(globalSettings);
  }
  let result = documentSettings.get(resource?.toString() ?? '');
  if (!result) {
    result = normalizeSettings(await connection.workspace.getConfiguration({
      scopeUri: resource?.toString(),
      section: 'VhdlLinter'
    }) as ISettings);
  }
  documentSettings.set(resource?.toString() ?? '', result);
  return result;
}

export function changeConfigurationHandler(change: { settings: { VhdlLinter?: ISettings } }) {

  if (currentCapabilities.configuration) {
    // Reset all cached document settings
    documentSettings.clear();
  } else {
    globalSettings = (
      (change.settings.VhdlLinter ?? defaultSettings)
    );
  }

  // Revalidate all open text documents
  for (const document of documents.all()) {
    void validateTextDocument(document);
  }
}