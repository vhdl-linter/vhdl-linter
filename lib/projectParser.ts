import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { existsSync, promises } from 'fs';
import { realpath } from 'fs/promises';
import { minimatch } from 'minimatch';
import { basename, dirname, join, sep } from 'path';
import { fileURLToPath, pathToFileURL } from 'url';
import { DeepPartial } from 'utility-types';
import { CancellationToken, Diagnostic, DiagnosticSeverity, RemoteWorkspace, WorkDoneProgressReporter } from 'vscode-languageserver';
import { Elaborate } from './elaborate/elaborate';
import { elaborateTargetLibrary } from './elaborate/elaborateTargetLibrary';
import { SetAdd } from './languageFeatures/findReferencesHandler';
import { OArchitecture, OConfigurationDeclaration, OContext, OEntity, OI, OPackage, OPackageInstantiation, ParserError } from './parser/objects';
import { ISettings, defaultSettings, settingsSchema } from './settingsGenerated';
import { VerilogParser } from './verilogParser';
import { VhdlLinter } from './vhdlLinter';
import { parse } from 'yaml';
import Ajv from "ajv";
import { currentCapabilities, trimSpacesOfStyleSettings, overwriteSettings } from './settingsUtil';

export function joinURL(url: URL, ...additional: string[]) {
  const path = join(fileURLToPath(url), ...additional);

  return pathToFileURL(path);
}
export function getRootDirectory() {
  let currentDir = pathToFileURL(__dirname);
  let iterations = 10;
  while (!existsSync(joinURL(currentDir, 'package.json'))) {
    currentDir = joinURL(currentDir, '..');
    if (iterations-- === 0) {
      throw new Error('Could not find root directory');
    }
  }
  return currentDir;
}
interface LibraryMapping {
  library: string;
  definitionFile: URL; // the file which defines the mapping
  definitionLine: number;
}

export function matchGlobList(value: string, globList: string[]) {
  return globList.some(glob => minimatch(basename(value), glob));
}

export const vhdlGlob = '*.vhd?(l)';
const verilogGlob = '*.?(s)v';
export const settingsGlob = 'vhdl-linter.yml';

export class ProjectParser {
  public cachedSettings: FileCacheSettings[] = [];
  public cachedFiles: (FileCacheVhdl | FileCacheVerilog | FileCacheLibraryList)[] = [];
  // maps files (url.toString()) to a library mapping
  public libraryMap = new Map<string, LibraryMapping>();
  public packages: OPackage[] = [];
  public packageInstantiations: OPackageInstantiation[] = [];
  public contexts: OContext[] = [];
  public entities: OEntity[] = [];
  public configurations: OConfigurationDeclaration[] = [];
  public architectures: OArchitecture[] = [];
  events = new EventEmitter();
  private watchers: FSWatcher[] = [];
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(workspaces: URL[], disableWatching = false,
    progress?: WorkDoneProgressReporter, vsCodeWorkspace?: RemoteWorkspace) {
    const projectParser = new ProjectParser(workspaces, progress, vsCodeWorkspace);
    await projectParser.init(disableWatching);
    return projectParser;
  }
  private constructor(public workspaces: URL[], public progress?: WorkDoneProgressReporter, public vsCodeWorkspace?: RemoteWorkspace) { }
  public async watchFolders(urls: URL[]) {
    for (const url of urls) {
      const settings = await this.getDocumentSettings(url);
      // Chokidar does not accept win style line endings
      // Chokidar sometimes throws this error message on hitting weird symlink. But this is a upstream issue
      // [Error: EISDIR: illegal operation on a directory, read] {
      //   errno: -21,
      //     code: 'EISDIR',
      //       syscall: 'read'
      // }
      const watcher = watch([
        fileURLToPath(url).replaceAll(sep, '/') + `/**/${vhdlGlob}`,
        fileURLToPath(url).replaceAll(sep, '/') + `/**/${settingsGlob}`,
        ...settings.analysis.verilogAnalysis ? [fileURLToPath(url).replaceAll(sep, '/') + `/**/${verilogGlob}`] : [],
        ...settings.paths.libraryMapFiles.map(glob => fileURLToPath(url).replaceAll(sep, '/') + `/**/${glob}`),
      ], { ignoreInitial: true, followSymlinks: false });
      watcher.on('add', (path) => {
        const handleEvent = async () => {
          const url = pathToFileURL(path);
          const settings = await this.getDocumentSettings(url);
          if (matchGlobList(path, [verilogGlob]) && settings.analysis.verilogAnalysis) {
            const cachedFile = await FileCacheVerilog.create(url, this, false);
            this.cachedFiles.push(cachedFile);
          } else if (matchGlobList(path, settings.paths.libraryMapFiles)) {
            const libraryFile = await FileCacheLibraryList.create(url, this);
            this.cachedFiles.push(libraryFile);
          } else if (matchGlobList(path, [settingsGlob])) {
            const settingsFile = await FileCacheSettings.create(url, this);
            this.cachedSettings.push(settingsFile);
          } else {
            const cachedFile = await FileCacheVhdl.create(url, this, false);
            this.cachedFiles.push(cachedFile);
          }
          this.flattenProject();
          this.events.emit('change', 'add', url.toString());
        };
        handleEvent().catch(console.error);
      });
      watcher.on('change', (path) => {
        const handleEvent = async () => {

          const url = pathToFileURL(path);
          if (matchGlobList(path, [settingsGlob])) {
            const cachedSetting = process.platform === 'win32'
              ? this.cachedSettings.find(cachedFile => cachedFile.uri.toString().toLowerCase() === url.toString().toLowerCase())
              : this.cachedSettings.find(cachedFile => cachedFile.uri.toString() === url.toString());
            await cachedSetting?.parse();
            this.events.emit('change', 'change', url.toString());
          } else {
            const cachedFile = process.platform === 'win32'
              ? this.cachedFiles.find(cachedFile => cachedFile.uri.toString().toLowerCase() === url.toString().toLowerCase())
              : this.cachedFiles.find(cachedFile => cachedFile.uri.toString() === url.toString());
            if (cachedFile) {
              await cachedFile.parse();
              this.flattenProject();
              this.events.emit('change', 'change', url.toString());
            } else {
              console.error('modified file not found', path);
            }
            this.cachedElaborate = undefined;
          }
        };
        handleEvent().catch(console.error);
      });
      watcher.on('unlink', path => {
        const url = pathToFileURL(path);

        if (matchGlobList(path, [settingsGlob])) {
          const cachedSettingIndex = process.platform === 'win32'
            ? this.cachedSettings.findIndex(cachedFile => cachedFile.uri.toString().toLowerCase() === url.toString().toLowerCase())
            : this.cachedSettings.findIndex(cachedFile => cachedFile.uri.toString() === url.toString());
          if (cachedSettingIndex > -1) {
            this.cachedSettings.splice(cachedSettingIndex, 1);
            this.events.emit('change', 'change', url.toString());
          }
        } else {
          const cachedFileIndex = process.platform === 'win32'
            ? this.cachedFiles.findIndex(cachedFile => cachedFile.uri.toString().toLowerCase() === url.toString().toLowerCase())
            : this.cachedFiles.findIndex(cachedFile => cachedFile.uri.toString() === url.toString());
          if (cachedFileIndex > -1) {
            this.cachedFiles.splice(cachedFileIndex, 1);
            this.flattenProject();
            this.events.emit('change', 'unlink', url.toString());
          }
        }
      });
      this.watchers.push(watcher);
    }
  }

  private parsedSettingsDirectories = new Set<string>();
  private async parseAllSettings(directory: URL) {
    const entries = await promises.readdir(directory);
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = joinURL(directory, entry);
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile() && matchGlobList(basename(entry), [settingsGlob])) {
          const settingsFile = await FileCacheSettings.create(filePath, this);
          this.cachedSettings.push(settingsFile);
        } else if (fileStat.isDirectory()) {
          const realPath = await realpath(filePath);
          // catch infinite recursion in symlink
          if (this.parsedSettingsDirectories.has(realPath) === false) {
            this.parsedSettingsDirectories.add(await realpath(realPath));
            await this.parseAllSettings(filePath);
          }
        }
      } catch (e) {
        console.error(e);
      }
    }));
  }

  private async init(disableWatching: boolean) {
    const files = new SetAdd<string>();
    // parse all settings files first
    await Promise.all(this.workspaces.map(async (directory) => {
      await this.parseAllSettings(directory);
    }));
    await Promise.all(this.workspaces.map(async (directory) => {
      const settings = await this.getDocumentSettings(directory);
      const directories = await this.parseDirectory(directory, settings.analysis.verilogAnalysis);
      files.add(...directories.map(url => url.toString()));
    }));
    const rootDirectory = getRootDirectory();
    const builtinFiles = (await this.parseDirectory(joinURL(rootDirectory, 'ieee2008'), false)).map(url => url.toString());
    files.add(...builtinFiles);
    let index = 0;
    for (const file of files) {
      if (this.progress) {
        const percent = Math.round(index / files.size * 100);
        this.progress.report(index / files.size * 100, `ProjectParser ${percent}% current file: ${basename(fileURLToPath(file))}`);
        index++;
      }
      const builtIn = builtinFiles.includes(file);
      const settings = await this.getDocumentSettings(new URL(file));
      // there should be no settings files here as they have been parsed in their own pass
      if (minimatch(basename(file), verilogGlob) && settings.analysis.verilogAnalysis) {
        const cachedFile = await FileCacheVerilog.create(new URL(file), this, builtIn);
        this.cachedFiles.push(cachedFile);
      } else if (minimatch(basename(file), vhdlGlob)) {
        const cachedFile = await FileCacheVhdl.create(new URL(file), this, builtIn);
        this.cachedFiles.push(cachedFile);
      } else if (matchGlobList(file, settings.paths.libraryMapFiles)) {
        const libraryFile = await FileCacheLibraryList.create(new URL(file), this);
        this.cachedFiles.push(libraryFile);
      }
    }
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    this.flattenProject();
    if (!disableWatching) {
      await this.watchFolders(this.workspaces);
    }
  }
  async stop() {
    for (const watcher of this.watchers) {
      await watcher.close();
    }
  }
  private parsedDirectories = new Set<string>();
  private async parseDirectory(directory: URL, parseVerilog: boolean): Promise<URL[]> {
    const files: URL[] = [];
    const settings = await this.getDocumentSettings(directory);
    const entries = await promises.readdir(directory);
    const ignoreRegex = settings.paths.ignoreRegex.trim().length > 0 ? new RegExp(settings.paths.ignoreRegex) : null;
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = joinURL(directory, entry);
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (matchGlobList(basename(entry), [vhdlGlob, verilogGlob, ...settings.paths.libraryMapFiles])
            && (matchGlobList(basename(entry), settings.paths.ignoreFiles) === false) && (ignoreRegex === null || !filePath.pathname.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else if (fileStat.isDirectory()) {
          const realPath = await realpath(filePath);
          // catch infinite recursion in symlink
          if (this.parsedDirectories.has(realPath) === false) {
            this.parsedDirectories.add(await realpath(realPath));
            files.push(... await this.parseDirectory(filePath, parseVerilog));
          }
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  relativeToWorkspace(url: URL) {
    const path = fileURLToPath(url);
    for (const workspace of this.workspaces) {
      if (path.startsWith(fileURLToPath(workspace))) {
        return path.replace(fileURLToPath(workspace), '').replace(sep, '/');
      }
    }
    return path.replace(sep, '/');
  }
  flattenProject() {
    this.entities = [];
    this.architectures = [];
    this.configurations = [];
    this.packages = [];
    this.packageInstantiations = [];
    this.contexts = [];
    this.libraryMap.clear();

    // do the library mapping at the beginning such that the library elaboration can happen at this stage
    for (const cachedFile of this.cachedFiles) {
      if (cachedFile instanceof FileCacheLibraryList) {
        for (const [file, lib] of cachedFile.libraryMap.entries()) {
          const existing = this.libraryMap.get(file);
          if (existing !== undefined && existing.library !== lib.library) {
            const message = {
              message: `This file has another and different library association: ${this.relativeToWorkspace(existing.definitionFile)}:${existing.definitionLine + 1} (library ${existing.library}).`,
              range: { start: { line: lib.definitionLine, character: 0 }, end: { line: lib.definitionLine, character: 1000 } },
              severity: DiagnosticSeverity.Warning,
            };
            // avoid duplicate messages
            if (cachedFile.messages.some(m => m.message === message.message && m.range.start.line === message.range.start.line) === false) {
              cachedFile.messages.push(message);
            }
            continue;
          }
          this.libraryMap.set(file, lib);
        }
      }
    }
    for (const cachedFile of this.cachedFiles) {

      if (cachedFile instanceof FileCacheVhdl) {
        elaborateTargetLibrary(cachedFile.linter);
        this.entities.push(...cachedFile.linter.file.entities);
        this.architectures.push(...cachedFile.linter.file.architectures);
        this.configurations.push(...cachedFile.linter.file.configurations);
        this.packages.push(...cachedFile.linter.file.packages.filter(pkg => pkg instanceof OPackage) as OPackage[]);
        this.packageInstantiations.push(...cachedFile.linter.file.packageInstantiations);
        this.contexts.push(...cachedFile.linter.file.contexts);
      } else if (cachedFile instanceof FileCacheVerilog) {
        this.entities.push(...cachedFile.parser.file.entities);
      }

    }
  }
  // Cache the elaboration result. Caution this can get invalid super easy. Therefore it is completely removed on any file change.
  private cachedElaborate: string | undefined = undefined;
  async elaborateAll(filter: string, token?: CancellationToken) {
    if (this.cachedElaborate === filter) {
      return;
    }
    const cachedFiles = this.cachedFiles.filter(cachedFile => cachedFile instanceof FileCacheVhdl && cachedFile.linter.file.lexerTokens.find(token => token.getLText() === filter)) as FileCacheVhdl[];
    for (const cachedFile of cachedFiles) {
      Elaborate.clear(cachedFile.linter);
    }

    for (const cachedFile of cachedFiles) {
      if (cachedFile.linter.file.lexerTokens.find(token => token.getLText() === filter)) {
        cachedFile.linter.token = token;
        await Elaborate.elaborate(cachedFile.linter);
      }
    }
    this.cachedElaborate = filter;

  }

  public findSettings(url: URL): FileCacheSettings | undefined {
    let path = fileURLToPath(url);
    // as long as we are not at the root
    while (dirname(path) !== path) {
      const res = this.cachedSettings.filter(cache => cache instanceof FileCacheSettings).find(cache => {
        const cachePath = fileURLToPath(cache.uri);
        const test = join(path, settingsGlob);
        return cachePath === test;
      });
      if (res) {
        return res;
      }
      path = dirname(path);
    }
  }

  // Cache the settings of all open documents
  public documentSettings = new Map<string, ISettings>();

  public async getDocumentSettings(resource: URL | undefined): Promise<ISettings> {
    // default settings are assumed as default and then overwritten by either
    // settings from vs code (workspace) or the closest vhdl-linter.yml
    if (resource !== undefined) {
      const fileSettings = this.findSettings(resource);
      if (fileSettings?.settings !== undefined) {
        return overwriteSettings(defaultSettings, fileSettings.settings);
      }
    }
    if (currentCapabilities.configuration === false) {
      return trimSpacesOfStyleSettings(defaultSettings);
    }
    let result = this.documentSettings.get(resource?.toString() ?? '');
    if (result === undefined && this.vsCodeWorkspace !== undefined) {
      result = trimSpacesOfStyleSettings(await this.vsCodeWorkspace.getConfiguration({
        scopeUri: resource?.toString(),
        section: 'VhdlLinter'
      }) as ISettings);
    }
    if (result === undefined) {
      result = trimSpacesOfStyleSettings(defaultSettings);
    }
    this.documentSettings.set(resource?.toString() ?? '', result);
    return result;
  }
}

export class FileCacheSettings {
  public messages: Diagnostic[] = [];
  public settings?: DeepPartial<ISettings>;
  lintingTime: number;
  public builtIn = false;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(uri: URL, projectParser: ProjectParser) {
    const cache = new FileCacheSettings(uri, projectParser);
    await cache.parse();
    return cache;
  }
  private constructor(public uri: URL, public projectParser: ProjectParser) {
  }
  async parse() {
    this.messages = [];
    this.settings = undefined;
    const text = await promises.readFile(this.uri, { encoding: 'utf8' });
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    let parsed = parse(text);
    if (parsed === null) {
      parsed = {};
    }
    const ajv = new Ajv();
    try {
      const validate = ajv.compile<DeepPartial<ISettings>>(settingsSchema);
      if (validate(parsed)) {
        this.settings = parsed;
      }
    } catch (e) {
      console.log(e);
    }
  }
}
export class FileCacheLibraryList {
  public messages: Diagnostic[] = [];
  public libraryMap = new Map<string, LibraryMapping>();
  lintingTime: number;
  public builtIn = false;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(uri: URL, projectParser: ProjectParser) {
    const cache = new FileCacheLibraryList(uri, projectParser);
    await cache.parse();
    return cache;
  }
  private constructor(public uri: URL, public projectParser: ProjectParser) {
  }
  async parse() {
    this.libraryMap.clear();
    this.messages = [];
    const text = await promises.readFile(this.uri, { encoding: 'utf8' });
    const lines = text.replaceAll('\r\n', '\n').split('\n');
    for (const [i, line] of lines.entries()) {
      if (line.trim() === '') {
        continue;
      }
      // expect `library,path`
      const split = line.split(',');
      if (split.length !== 2) {
        this.messages.push({
          message: `Expected library,path but got ${split.length} parts.`,
          range: { start: { line: i, character: 0 }, end: { line: i, character: line.length } },
          severity: DiagnosticSeverity.Warning,
        });
        continue;
      }
      if (split[1]!.startsWith('/')) {
        // absolute path
        this.messages.push({
          message: `Only relative paths are allowed.`,
          range: { start: { line: i, character: split[0]!.length + 1 }, end: { line: i, character: line.length } },
          severity: DiagnosticSeverity.Warning,
        });
        continue;
      }
      const path = pathToFileURL(dirname(fileURLToPath(this.uri)));
      const url = joinURL(path, split[1]!);
      if (existsSync(fileURLToPath(url)) === false) {
        this.messages.push({
          message: `${this.projectParser.relativeToWorkspace(url)} does not exist.`,
          range: { start: { line: i, character: split[0]!.length + 1 }, end: { line: i, character: line.length } },
          severity: DiagnosticSeverity.Warning,
        });
      }
      const existing = this.libraryMap.get(url.toString());
      if (existing !== undefined) {
        this.messages.push({
          message: `This file already has a library associated with it (${existing.library} in line ${existing.definitionLine + 1}).`,
          range: { start: { line: i, character: 0 }, end: { line: i, character: line.length } },
          severity: DiagnosticSeverity.Warning,
        });
        continue;
      }
      this.libraryMap.set(url.toString(), {
        library: split[0]!,
        definitionFile: this.uri,
        definitionLine: i
      });
    }
  }

}
export class FileCacheVhdl {
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  linter: VhdlLinter;
  lintingTime: number;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(uri: URL, projectParser: ProjectParser, builtIn: boolean) {
    const cache = new FileCacheVhdl(uri, projectParser, builtIn);
    await cache.parse();
    return cache;
  }
  private constructor(public uri: URL, public projectParser: ProjectParser, public builtIn: boolean) {
  }
  async parse() {
    const stat = await promises.stat(this.uri);
    let text;
    const maxFileSize = (await this.projectParser.getDocumentSettings(this.uri)).analysis.maxFileSize;
    if (stat.size > maxFileSize) {
      text = '';
    } else {
      text = await promises.readFile(this.uri, { encoding: 'utf8' });
      text = text.replaceAll('\r\n', '\n');
    }
    this.linter = new VhdlLinter(this.uri, text, this.projectParser, await this.projectParser.getDocumentSettings(this.uri));
    this.replaceLinter(this.linter);
  }
  replaceLinter(vhdlLinter: VhdlLinter) {
    this.linter = vhdlLinter;
  }

}
class FileCacheVerilog {
  lintingTime: number;
  parser: VerilogParser;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(uri: URL, projectParser: ProjectParser, builtIn: boolean) {
    const cache = new FileCacheVerilog(uri, projectParser, builtIn);
    await cache.parse();
    return cache;
  }
  private constructor(public uri: URL, public projectParser: ProjectParser, public builtIn: boolean) {
  }
  async parse() {
    const stat = await promises.stat(this.uri);
    let text;
    const maxFileSize = (await this.projectParser.getDocumentSettings(this.uri)).analysis.maxFileSize;
    if (stat.size > maxFileSize) {
      text = '';
    } else {
      text = await promises.readFile(this.uri, { encoding: 'utf8' });
      text = text.replaceAll('\r\n', '\n');
    }
    this.parser = new VerilogParser(this.uri, text, this.projectParser);
  }

}