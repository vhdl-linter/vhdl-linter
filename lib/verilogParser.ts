import { OLexerToken, TokenType } from "./lexer";
import * as O from "./parser/objects";
import { OIRange } from "./parser/objects";
import { OEntity, OFile } from "./parser/objects";
import { ProjectParser } from "./projectParser";
import { SettingsGetter } from "./vhdlLinter";

export class VerilogParser {
  file: OFile;
  pos = 0;
  advanceWhitespace() {
    const match = this.text.substring(this.pos).match(/^\s+/);
    if (match) {
      this.pos += match[0].length;
    }
  }
  parsePortOrParameter(parent: O.OEntity, str: string, offset: number, parameter: boolean) {
    const regex = /^(\s*(input|inout|output|parameter)?\s+(supply0|supply1|tri|triand|trior|tri0|tri1|uwire|wire|wand|wor|reg|logic|longint)?\s*(?:\[.*?\])?\s*)([a-zA-Z]\w*)\s*(?:\[.*?\])?\s*(?:=\s*(\S+))?/s;
    // groups:
    // 1: everything until name
    // 2 (optional): in, out, parameter etc.
    // 3 (optional): net_type
    // 4: name
    // 5 (optional): default value
    const match = str.match(regex);
    if (match) {
      let port: O.OGenericConstant | O.OPort;
      if (parameter) {
        port = new O.OGenericConstant(parent, new O.OIRange(this.file, offset, offset + str.length));
      } else {
        port = new O.OPort(parent, new O.OIRange(this.file, offset, offset + str.length));
        if (match[2] === 'output') {
          port.direction = 'out';
        } else if (match[2] === 'inout') {
          port.direction = 'inout';
        } else { // default to input
          port.direction = 'in';
        }
      }
      port.lexerToken = new OLexerToken(match[4]!, new O.OIRange(this.file, offset + match[1]!.length, offset + match[1]!.length + match[4]!.length), TokenType.implicit, this.file);
      if (match[5] !== undefined) {
        port.defaultValue = [new O.OName(port, new OLexerToken(match[5].replace(/=\s*/s, ''), new OIRange(this.file, 0, 0), TokenType.implicit, this.file))];
      }
      if (port instanceof O.OGenericConstant) {
        parent.generics.push(port);
      } else {
        parent.ports.push(port);
      }
    }
  }

  constructor(public uri: URL, public text: string, public projectParser: ProjectParser,
    public settingsGetter: SettingsGetter,
  ) {
    const originalText = this.text;

    // remove multi line and single line comments
    this.text = this.text.replaceAll(/\/\*(.*?)\*\//sg, match => match.replaceAll(/[^\n]/g, ' '));
    this.text = this.text.replaceAll(/\/\/.*/g, match => ' '.repeat(match.length));
    let match;
    this.file = new OFile(this.text, this.uri, originalText, []);
    let iterationCounter = 100;
    // find modules
    // eslint-disable-next-line no-cond-assign
    while (match = this.text.substring(this.pos).match(/\b(module\s+)(\w+)/s)) {
      if ((iterationCounter-- === 0)) {
        console.log(`Infinite loop in ${uri.toString()}`);
        return;
      }
      const moduleName = new OLexerToken(match[2]!, new O.OIRange(this.file, this.pos + match.index! + match[1]!.length, this.pos + match.index! + match[2]!.length), TokenType.implicit, this.file);
      const module = new OEntity(this.file, new O.OIRange(this.file, this.pos + match.index!, 0));
      this.pos += match.index! + match[0].length;
      this.advanceWhitespace();
      // find parameters, i.e. #(...)
      // eslint-disable-next-line no-cond-assign
      if (match = this.text.substring(this.pos).match(/#\((.*?)\)/s)) {
        let offset = this.pos + match.index! + 2;
        this.pos += match.index! + match[0].length;
        const parametersString = match[1]!;
        for (const parameterString of parametersString.split(',')) {
          this.parsePortOrParameter(module, parameterString, offset, true);
          offset += parameterString.length + 1;
        }
      }
      this.advanceWhitespace();
      // find port declaration, i.e. (...)
      // eslint-disable-next-line no-cond-assign
      if (match = this.text.substring(this.pos).match(/\((.*?)\)/s)) {
        let offset = this.pos + match.index! + 1;
        this.pos += match.index! + match[0].length;
        const portsString = match[1]!;
        for (const portString of portsString.split(',')) {
          this.parsePortOrParameter(module, portString, offset, false);
          offset += portString.length + 1;
        }
      }
      match = this.text.substring(this.pos).match(/\bendmodule\b/);
      module.lexerToken = moduleName;
      if (match) {
        module.range = module.range.copyWithNewEnd(new O.OI(this.file, this.pos + match.index!));
        this.pos += match.index! + match[0].length;
      }

      this.file.entities.push(module);
    }
  }
}