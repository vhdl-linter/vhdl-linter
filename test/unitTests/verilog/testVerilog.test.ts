import { expect, test } from '@jest/globals';
import { join } from 'path';
import { pathToFileURL } from 'url';
import { ProjectParser } from '../../../lib/projectParser';
import { VerilogParser } from '../../../lib/verilogParser';
import { VhdlLinter } from '../../../lib/vhdlLinter';
import { makeRangePrintable } from '../../helper';
import { readFileSyncNorm } from '../../readFileSyncNorm';
import { getDocumentSettings } from '../../../lib/settingsManager';
import { rmSync, writeFileSync } from 'fs';


test.each([true, false])('Testing verilog switch %b', async enableVerilog => {
  const settingsFile = join(__dirname, 'vhdl-linter.yml');
  writeFileSync(settingsFile, JSON.stringify({ analysis: { verilogAnalysis: enableVerilog } }));
  const projectParser = await ProjectParser.create([pathToFileURL(__dirname)]);
  const uri = pathToFileURL(join(__dirname, '_test_verilog_instance.vhd'));
  const settings = await getDocumentSettings(uri, projectParser);
  const linter = new VhdlLinter(uri, readFileSyncNorm(uri, { encoding: 'utf8' }), projectParser, settings);
  await linter.checkAll();
  if (enableVerilog) {
    expect(linter.messages).toHaveLength(0);
  } else {
    expect(linter.messages).toHaveLength(1);
  }

  await projectParser.stop();
  rmSync(settingsFile);
});

test('module_advanced', async () => {
  const uri = pathToFileURL(join(__dirname, 'module_advanced.sv'));
  const projectParser = await ProjectParser.create([]);
  const verilogParser = new VerilogParser(uri, readFileSyncNorm(uri, { encoding: 'utf8' }), projectParser);
  expect(verilogParser.file.objectList.map(obj => ({
    range: makeRangePrintable(obj.range),
    lexerToken: obj.lexerToken?.text
  }))).toMatchSnapshot();
});