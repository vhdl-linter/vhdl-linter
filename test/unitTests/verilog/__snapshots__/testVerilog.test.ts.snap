// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module snapshot module_advanced.sv 1`] = `
[
  {
    "lexerToken": "switch",
    "range": "1:1 - 14:1",
  },
  {
    "lexerToken": "g_DEVICES",
    "range": "2:2 - 2:12",
  },
  {
    "lexerToken": "g_FREQUENCY",
    "range": "3:2 - 3:14",
  },
  {
    "lexerToken": "g_AVALON_LENGTH",
    "range": "4:2 - 4:22",
  },
  {
    "lexerToken": undefined,
    "range": "1:1 - 1:1",
  },
  {
    "lexerToken": "annoying",
    "range": "5:3 - 5:69",
  },
  {
    "lexerToken": undefined,
    "range": "1:1 - 1:1",
  },
  {
    "lexerToken": "g_TEST",
    "range": "6:3 - 7:1",
  },
  {
    "lexerToken": undefined,
    "range": "1:1 - 1:1",
  },
  {
    "lexerToken": "i_clk",
    "range": "8:2 - 8:8",
  },
  {
    "lexerToken": "i_portSpeed",
    "range": "9:3 - 9:47",
  },
  {
    "lexerToken": "o_txReady",
    "range": "10:3 - 10:44",
  },
  {
    "lexerToken": "i_avalonVec",
    "range": "11:3 - 11:62",
  },
  {
    "lexerToken": "o_avalonVec",
    "range": "12:3 - 13:1",
  },
]
`;

exports[`test verilog with non ansi header 1`] = `
[
  {
    "lexerToken": "dummy_module_non_ansi",
    "range": "3:1 - 32:1",
  },
  {
    "lexerToken": "A",
    "range": "5:3 - 5:5",
  },
  {
    "lexerToken": "B",
    "range": "6:3 - 6:5",
  },
  {
    "lexerToken": "C",
    "range": "8:3 - 10:4",
  },
  {
    "lexerToken": "PARAM_A",
    "range": "13:3 - 13:25",
  },
  {
    "lexerToken": undefined,
    "range": "1:1 - 1:1",
  },
  {
    "lexerToken": "PARAM_B",
    "range": "15:3 - 15:20",
  },
  {
    "lexerToken": "PARAM_C",
    "range": "15:21 - 15:34",
  },
  {
    "lexerToken": undefined,
    "range": "1:1 - 1:1",
  },
]
`;
